bgpv2: Adding preflight and neighbor reconcilers using CiliumBGPNodeConfig CRD.

harsimran-pabla opened the pull request:
This change introduces scaffolding for BGPv2 reconcilers, starting with preflight and neighbor reconcilers. At later stage, other reconcilers will be added. 

Change includes two commits
- First commit introduces preflight and neighbor reconcilers. Along with the unit tests. 
- Second commit adds ReconcileInstances API to BGPRouteManager. Along with all the changes required to manage BGP instances.

```release-note
bgpv2 - adding preflight and neighbor reconciler using CiliumBGPNodeConfig resource.
```

rastislavs commented on pkg/bgpv1/manager/instance/instance.go:
```diff
@@ -55,3 +55,31 @@ func NewServerWithConfig(ctx context.Context, log *logrus.Entry, params types.Se
 }
+
+// BGPInstance is a container for providing interface with underlying router implementation.
+type BGPInstance struct {
```
I would note here that `ServerWithConfig` is used for BGPv1 and `BGPInstance` for BGPv2

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -82,46 +87,27 @@ type BGPRouterManager struct {
 
 	// running is set when the manager is running, and unset when it is stopped.
 	running bool
+
```
I would also mark `Servers` and `Reconcilers` as BGPv1

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -502,6 +488,276 @@ func (m *BGPRouterManager) Stop() {
+func (m *BGPRouterManager) registerV2(ctx context.Context, rd *reconcileDiffV2) error {
+	l := log.WithFields(
+		logrus.Fields{
+			"component": "manager.v2.add",
```
"manager.v2.register"?

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -502,6 +488,276 @@ func (m *BGPRouterManager) Stop() {
+			continue
+		}
+		if err := m.registerBGPInstance(ctx, config, rd.ciliumNode); err != nil {
+			// we'll just log the error and attempt to register the next BgpServer.
```
Why don't we return an error here instead? Wouldn't it be problematic later to introduce a retry logic if we just skip the error?

harsimran-pabla replied:
yes, you're right. Later if we introduce retry logic - we will not push up the error. I think original idea was that each instance is unique and failure to register one should not impact registration of second instance. 

I think we can still continue with this approach, but push out an error if any instance fails to register. That way, retry logic will catch the error and try reconciliation again.

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -502,6 +488,276 @@ func (m *BGPRouterManager) Stop() {
+func (m *BGPRouterManager) reconcileV2(ctx context.Context, rd *reconcileDiffV2) error {
+	l := log.WithFields(
+		logrus.Fields{
+			"component": "manager.reconcile",
```
"manager.v2.reconcile"?

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+	var (
+		l = log.WithFields(
+			logrus.Fields{
+				"component": "reconciler.neighbor",
```
maybe we could use `"reconcilerv2."+r.Name()` as the component name for all reconcilers?

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/preflight.go:
```diff
@@ -0,0 +1,196 @@
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+// PreflightReconciler reconciles BPG Global configuration. This reconciler is similar to preflight reconciler.
```
```suggestion
// PreflightReconciler reconciles BPG Global configuration. This reconciler is similar to v1 preflight reconciler.
```

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/preflight.go:
```diff
@@ -0,0 +1,196 @@
+	// stop the old BgpServer
+	p.BGPInstance.Router.Stop()
+
+	// create a new one via ServerWithConfig constructor
```
is `ServerWithConfig` really used here?

rastislavs commented on pkg/bgpv1/manager/workdiff.go:
```diff
@@ -112,3 +113,90 @@ func (wd *reconcileDiff) withdrawDiff(m LocalASNMap) error {
+//
+// since registerOrReconcileDiff populates the `seen` field of a diff, this method should always
+// be called first when computing a reconcileDiff.
+func (wd *reconcileDiffV2) registerOrReconcileDiff(m map[string]*instance.BGPInstance, nodeObj *v2alpha1api.CiliumBGPNodeConfig) error {
```
Maybe it's just me, but I find it not easy to read the diff logic - I think it could be easier to read with some changes in the variable naming in `diff`, `registerOrReconcileDiff` and `withdrawDiff` methods:

- `m` to something like `existing` or `existingInstances`
- `nodeObj` to something like `desired` or `desiredConfig`

rastislavs reviewed (commented):
Thanks, I have some non-critical comments, but overall the approach makes sense to me.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+	Password string
+}
+
+func (p *PeerData) Equal(other *PeerData) bool {
```
Can't we generate this function by deep-equal-gen? Otherwise, it's hard to maintain the change to the PeerData.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+			logrus.Fields{
+				"component": fmt.Sprintf("reconciler.v2.%s", r.Name()),
+			},
+		)
```
Let's don't do "component" logging in v2. I think we have two problems.

1. Users usually index log fields into their log store (e.g., elastic search) and search for specific fields. "component" field exposes implementation details to users. It is subject to change at anytime, so it's basically the query using the "component" field is useless from the user's perspective. IMO, we should consider the log field as a user interface we need to maintain (at least for error/warning/info logging, debug logging could be more flexible, I think).
2. It's hard to maintain for us as well. We usually hard-code function name as a "component" I saw some of them are already out of sync. For example, we can see it in this PR as well https://github.com/cilium/cilium/pull/30108/commits/e0622fffe6f7994cedd337a26ca8cb126ce0491c#diff-c06abd174799f2fedd8fc17f5fcfe231ee52a690a45c9d5d4b2541774260f012R323-R327)

harsimran-pabla replied:
Would it make sense to have bgpv1/log package ? Moving `logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-control-plane")` to this package. BGP sub-system can use this as common logger.

harsimran-pabla replied:
Removed all component tags from logger. Now most of the logging k/v are BGP instance name etc, which is more meaningful to user and they can filter logs based on these keys.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+
+func (r *NeighborReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
+	if p.DesiredConfig == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil CiliumBGPNodeInstance")
```
Let's not depend on the name of the internal structure. It exposes implementation details to users through error logging.

Also, when this can happen? Do users need to take any action when they see it?

harsimran-pabla replied:
These errors are mostly for developers to debug problem. I don't think we should run into this condition, it is more of pragmatic approach, that if we have this corner case - we should error out instead of panic at later stage due to unexpected request. I think having internal details in error message is useful for doing RCA. From user prospective, if they see this error - they should probably just open an issue.

YutaroHayakawa replied:
Ok, then how about prefixing the error with `BUG:`? So that we can encourage users to file an issue. Same for below.

harsimran-pabla replied:
My concern with adding 'bug' in error message is it's not scalable. Most of the errors are a problem, most likely a bug. Logging error imo itself an indication of a problem.

YutaroHayakawa replied:
> Most of the errors are a problem, most likely a bug

Error and bug are two different things IMO.

For example, [here](https://github.com/cilium/cilium/pull/30108/files#diff-23b500b5d2af6a22a607b920886922d961deb0e5f8c686f7481e4320722e017fR646) we return an error when the node annotation has a format error. We log it as an Error later. However, this indicates the problem of the user's configuration and not a bug. Our code has no problem, but the user's configuration has a problem. This is an example of `Error`.

On the other hand, in this case, our code is likely to have a bug when we hit this condition. Users have nothing to do; we know the only way to solve this is by modifying the code. Then, we want them to report an issue. This is an example of `Bug`.

Both cases are currently reported as an error-level log. Thus, users don't have any way to distinguish those two. That's why I'm suggesting to indicate it is a `BUG` explicitly.

rastislavs replied:
I can see the `BUG:` prefix pattern in some other cilium packages as well, so in general I wouldn't be opposed it.
To me, it makes sense in logs, but it appears a bit strange in error strings - which often can be wrapped in another error, so the `BUG:` prefix may eventually appear in the middle of the log line when eventually logged.

What about logging with `BUG:` prefix at this place, and returning an error without the prefix?

harsimran-pabla replied:
In these particular cases, we are logging error message so I just prefixed the error message with `BUG:`. I am not fully opposed to the idea as it can be context dependent and maybe help users to raise issue by looking at the errors. 

( feel free to ignore rest of the comment )

My counter argument is 'error' level log is enough to warrant investigation. If error is result of a bug or incorrect configuration, it is impossible to tell without looking more into the context around it. 

In the example of node annotation - User sending incorrectly formatted annotation is user config issue, however if we introduced a bug by modifying annotation parsing, it is not a user error. In which case, we have to investigate the logged error message. 

From logging prospective, the obvious cases which should never ever happen should result in fatal. But that is bit extreme as well. 

I guess, main intention here is to get attention of the user by logging 'BUG' in the error message. My main point is that every error message should get a red-flag and warrants the investigation. Initially by user to validate that configuration is correct and then an issue to get extra eyes on it.

Anyway, it's an interesting thought experiment. I prefixed the logs with BUG in this case :)

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+		return fmt.Errorf("attempted neighbor reconciliation with nil CiliumBGPNodeInstance")
+	}
+	if p.BGPInstance == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil BGPInstance")
```
Same above.

harsimran-pabla replied:
resolving this comment, we can discuss in above if this need to be modified.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+	}
+
+	if len(toCreate) > 0 || len(toRemove) > 0 || len(toUpdate) > 0 {
+		l.Infof("Reconciling peers for virtual router with %s", p.DesiredConfig.Name)
```
We shouldn't use format strings in info logging. Users should be able to filter logs by fields. Same for all other places.

YutaroHayakawa replied:
Maybe, later we should expose metrics for this?

harsimran-pabla replied:
I am fixing all logs to have k/v pairs instead of string format. 
Yes, having metrics hooks into reconcilers will be useful. Each reconciler can expose appropriate statistics.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+		curNeigh []*PeerData = nil
+	)
+	newNeigh := p.DesiredConfig.Peers
+	l.Debugf("Begin reconciling peers for virtual router %s", p.DesiredConfig.Name)
```
This log is generated for every reconciliation. Let's consider making this a metric.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+			PeerConfig: n.Config,
+			Password:   n.Password,
+		}); err != nil {
+			return fmt.Errorf("failed while reconciling neighbor  %s: %w", n.Peer.Name, err)
```
Can we be more specific? Like "failed to update neighbor configuration to router"?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+
+	// create new neighbors
+	for _, n := range toCreate {
+		l.Infof("Adding peer %s to router %s", n.Peer.Name, p.DesiredConfig.Name)
```
Maybe, later we should expose metrics for create/delete/update?

harsimran-pabla replied:
Right, we need to figure out cardinality of these metrics - here it is number of instances - This is number is not expected to be large but it is not under our direct control. Under stable conditions counter for these metrics is going to be constant number.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+		if err := p.BGPInstance.Router.RemoveNeighbor(ctx, types.NeighborRequest{
+			Peer: n.Peer,
+		}); err != nil {
+			return fmt.Errorf("failed while reconciling neighbor  %s: %w", n.Peer.Name, err)
```
Can we be more specific? Like "failed to delete neighbor configuration from router"?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+		r.deleteMetadata(p.BGPInstance, p.DesiredConfig.Name, n)
+	}
+
+	l.Infof("Done reconciling peers for virtual router %s", p.DesiredConfig.Name)
```
This log message is generated for every reconciliation. Let's downgrade it to debug or expose metrics.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/preflight.go:
```diff
@@ -0,0 +1,196 @@
+		if attrs.LocalPort != 0 {
+			localPort = attrs.LocalPort
+		}
+	}
```
Do we still use annotations? I thought we were replacing it with `CiliumBGPNodeConfigOverride`.

harsimran-pabla replied:
As discussed offline, annotations are still used for backwards compatibility. If user sets node level configuration via CiliumBGPNodeConfigOverride - it will take precedence over annotations.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/preflight.go:
```diff
@@ -0,0 +1,196 @@
+	// create a new one via BGPInstance constructor
+	s, err := instance.NewBGPInstance(ctx, log, globalConfig)
+	if err != nil {
+		l.WithError(err).Errorf("Failed to start BGP server for virtual router %s", p.DesiredConfig.Name)
```
Do we need this? We will log the below error later, right?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/preflight_test.go:
```diff
@@ -0,0 +1,285 @@
+// TestPreflightReconciler ensures if a BgpServer must be recreated, due to
+// permanent configuration of the said server changing, its done so correctly.
+func TestPreflightReconciler(t *testing.T) {
+	req := require.New(t)
```
Nit: Is this something recommended for now? We used to use global functions of the required package, right?

YutaroHayakawa replied:
Ah, got it. We can omit the first `t` argument with this. Good tip...

BTW, don't we need to create it per `t.Run` ?

harsimran-pabla replied:
I don't think it is required ( atleast I never encountered any problem with this approach ). This is basically assertion, so it stops the test when failure is encountered instead of continuing with other test cases. I find this easier so failed test case is obvious.

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -502,6 +494,278 @@ func (m *BGPRouterManager) Stop() {
+		var config *v2alpha1api.CiliumBGPNodeInstance
+		var ok bool
+		if config, ok = rd.seen[name]; !ok {
+			l.Errorf("Work diff (add) contains unseen instance %s, skipping", name)
```
When does this happen? What should users do when they see this error?

harsimran-pabla replied:
This is same code as before - this is unlikely to happen as seen map is populated earlier in the call stack. There are no concurrent writes or reads to this map. 

From user prospective, error is not actionable - just opening an issue with the logs so we can investigate why it might have happened.

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -502,6 +494,278 @@ func (m *BGPRouterManager) Stop() {
+		} else {
+			routerID = nodeIP.String()
+		}
+	}
```
Do we still need annotation?

harsimran-pabla replied:
Yes, as discussed above.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,362 @@
+
+	// set defaults for config
+	config.Spec.SetDefaults()
+	return &config.Spec, nil
```
```suggestion
        var conf *v2alpha1.CiliumBGPPeerConfigSpec
	if !exists {
		// if config does not exist, return default config
		conf = &v2alpha1.CiliumBGPPeerConfigSpec{}
	} else {
	        conf = config.Spec
	}
	
        conf.SetDefaults()
        
	return config, nil
```

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/preflight.go:
```diff
@@ -0,0 +1,196 @@
+	annoMap, err := agent.NewAnnotationMap(p.CiliumNode.Annotations)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse Node annotations for virtual router %v: %w", p.DesiredConfig.Name, err)
+	}
```
Do we still use annotations? I thought we were replacing it with CiliumBGPNodeConfigOverride.

harsimran-pabla replied:
yes, still required for backwards compatability.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/preflight.go:
```diff
@@ -0,0 +1,196 @@
+		return "", fmt.Errorf("failed to parse Node annotations for virtual router %v: %w", p.DesiredConfig.Name, err)
+	}
+
+	routerID, err := annoMap.ResolveRouterID(asn)
+	if err != nil {
+		if nodeIP := p.CiliumNode.GetIP(false); nodeIP == nil {
+			return "", fmt.Errorf("failed to get ciliumnode IP %v: %w", nodeIP, err)
+		} else {
+			routerID = nodeIP.String()
+		}
+	}
```
Let's add a test for this auto detection case.

harsimran-pabla replied:
Added a test case to pick router ID from node IP which is derived from CiliumNode object.

YutaroHayakawa reviewed (changes requested):
Sorry for the delay 🙇 

Made an initial review. Thanks for working on this!

harsimran-pabla committed:
bgpv2: add new reconcilerV2 scaffolding and base reconcilers.

Introducing preflight and neighbor reconciler using CiliumBGPNodeConfig
and CiliumBGPPeerConfig resources. Reconcilers are added in reconcilerv2
pkg.

Preflight reconciler is used to create, delete and update BGP instances.
It differs from BGPv1 preflight reconciler. It overrides RouterID and
listening port from CiliumBGPNodeConfig, if they are set.

Neighbor reconciler is used to create, delete and update BGP peer
configuration per instance. It differs from BGPv1 reconciler in terms to
metadata storage and how peer configuration is created. In case of v2
reconcilers, peer configuration is fetched from CiliumBGPPeerConfig CRD
resource.

These reconcilers will need to be registered with BGP cell, this will be
done in future changes.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: manager and workdiff changes to support CiliumBGPNodeConfig.

This change introduces new API ReconcileInstances to BGPRouteManager.
This API will be called by controller when BGPv2 is enabled and
reconciliation is triggered.

New BGPv2 equivalent methods are added to workdiff and BGPRouteManager.
This is done because new configuration is based on CiliumBGPNodeConfig
instead of CiliumBGPPeeringPolicy.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

YutaroHayakawa commented on pkg/bgpv1/logger/log.go:
```diff
@@ -0,0 +1,35 @@
+var (
+	// Log to be used in BGP package. All log messages emitted by this have k/v
+	// `subsys=bgp-control-plane`.
+	Log = logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-control-plane")
```
You can DI this logger through Hive. You can inject the type `logrus.FieldLogger`. The logger already contains the field `subsys=<cell module name>`.

harsimran-pabla replied:
Taking this approach - as all the reconcilers are cells -it's easy to inject logger this way. Will move log field KVs to types package.

YutaroHayakawa replied:
Yep, notice that the subsys is the name of the Module, so when we DI this to reconcilers, it should still have `subsys=bgp-cp`. I would add a field `reconciler=<pod-cidr|service|pod-ippool|etc...>` (maybe this tag should indicate "what" it is reconciling from the user's point of view) for each reconciler's logger to make filtering easier.

harsimran-pabla replied:
Added new reconciler log field.

YutaroHayakawa commented on pkg/bgpv1/logger/log.go:
```diff
@@ -0,0 +1,35 @@
+
+	// Peer is used as key for BGP peer in the log field.
+	Peer = "bgp_peer"
+)
```
When we have `subsys` tag, it is obvious for users that the log is from BGP, so it's redundant to prefix log fields with `bgp`?

harsimran-pabla replied:
ack - make sense

YutaroHayakawa reviewed (changes requested):
Did a second review. Thanks for the fix!

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -529,6 +524,260 @@ func (m *BGPRouterManager) Stop() {
+			}
+			delete(m.BGPInstances, c.Name) // optimistic delete
+		}
+	}()
```
Related to this PR: https://github.com/cilium/cilium/pull/30382

When we have an error in the initial reconciliation, this causes a create/delete (peer disconnection/route update/delete) loop of the BGP server. We can, instead, register the server right after `instance.NewBGPInstance`. Then, even if the reconciliation fails, we don't need to re-do server creation.

harsimran-pabla replied:
Latest change fixes it.

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -529,6 +524,260 @@ func (m *BGPRouterManager) Stop() {
+		delete(m.BGPInstances, name)
+		m.Logger.WithField(types.InstanceLogField, name).Info("Removed BGP instance")
+	}
+	return nil
```
Related: https://github.com/cilium/cilium/pull/30382

This function never returns an error. We can modify the signature to not return an error.

harsimran-pabla replied:
Latest change fixes it.

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -529,6 +524,260 @@ func (m *BGPRouterManager) Stop() {
+	for name := range m.BGPInstances {
+		rd.withdraw = append(rd.withdraw, name)
+	}
+	return m.withdrawV2(ctx, rd)
```
Same above

harsimran-pabla replied:
Latest change fixes it.

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -529,6 +524,260 @@ func (m *BGPRouterManager) Stop() {
+		if err := m.reconcileBGPConfigV2(ctx, i, newc, rd.ciliumNode); err != nil {
+			m.Logger.WithField(types.InstanceLogField, name).WithError(err).Error("Encountered error reconciling BGP instance, shutting down this server")
+			i.Router.Stop()
+			delete(m.BGPInstances, name)
```
Related: https://github.com/cilium/cilium/pull/30382

Do we want to solve this in this PR?

harsimran-pabla replied:
Latest change fixes it.

YutaroHayakawa reviewed (changes requested):
Did a third review. I added a comment for the issue I found while I was working for this PR https://github.com/cilium/cilium/pull/30382.

Since we derived new reconcilers from the v1 reconciler, it obviously contains these issues we had in the bgpv1. I started to wonder how much of the issue we want to solve in this PR.

One option is, we only fix "critical" issues and let rest of the fixes to go into the follow-up if we can do that later. The cons of this approach is it is very easy to deprioritize the follow-up once this feature starts to "work". In that sense, we should solve any concern in this PR. WDYT?

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -529,6 +523,256 @@ func (m *BGPRouterManager) Stop() {
+
+	annoMap, err := agent.NewAnnotationMap(ciliumNode.Annotations)
+	if err != nil {
+		return fmt.Errorf("unable to parse local node annotations: %v", err)
```
Nit: "%v" => "%w"?

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -529,6 +523,256 @@ func (m *BGPRouterManager) Stop() {
+		}
+		if rErr := m.registerBGPInstance(ctx, config, rd.ciliumNode); rErr != nil {
+			// we'll log the error and attempt to register the next instance.
+			m.Logger.WithField(types.InstanceLogField, name).WithError(rErr).Errorf("Error while registering new BGP instance")
```
Here, we log an error with the field logger with the instance name. So, we don't need to include the instance name in the error message returned from functions called from `registerBGPInstance`. I see some errors containing instance names. Same for `reconcileV2`.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/neighbor.go:
```diff
@@ -0,0 +1,359 @@
+	for _, n := range toCreate {
+		l.WithField(types.PeerLogField, n.Peer.Name).Info("Adding peer")
+
+		if err := p.BGPInstance.Router.AddNeighbor(ctx, types.NeighborRequest{
+			Peer:       n.Peer,
+			PeerConfig: n.Config,
+			Password:   n.Password,
+		}); err != nil {
+			return fmt.Errorf("failed to add neigbhor %s in instance %s: %w", n.Peer.Name, p.DesiredConfig.Name, err)
+		}
```
We can revisit this later, but should we give up reconciliation immediately when `AddNeighbor` fails? Same for Update/Remove.

harsimran-pabla replied:
I think we can revisit it later, it requires behavior change and more tests. Right now I wanted to keep behaviour similar to BGPv1.

YutaroHayakawa reviewed (changes requested):
I think this is my last review 🙏

YutaroHayakawa reviewed (approved):
Now, LGTM. Thanks!
